### PR TITLE
LUN-440: Select image from gallery

### PR DESCRIPTION
**Pull request by LeandraH**
This pull request belongs to the issue https://issues.tuerantuer.org/browse/LUN-440.

### CODE REVIEW

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 1]
```diff
@@ -33,39 +42,92 @@ const Container = styled.View`
   justify-content: center;
 `
 
+const Photo = styled.Image`
+  width: ${wp('32%')}px;
+  height: ${wp('32%')}px;
+  margin: ${wp('0.6%')}px;
```
Comment (f1sh1918) [inline src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:48]: please use our spacing classes for margin

**Review by f1sh1918 — CHANGES_REQUESTED** [review]
Tested on android real device.
The permission request for for media is not showing up instead the `NoAuthView` shows up at instant

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 2]
```diff
@@ -33,39 +42,92 @@ const Container = styled.View`
...
-  pushImage: (imageUri: string) => void
+  pushImage: (image: UserVocabularyImage) => void
 }
 
 const ImageSelectionOverlay = ({ setVisible, pushImage }: ImageSelectionOverlayProps): ReactElement => {
   const camera = useRef<RNCamera>(null)
+  const [showGallery, setShowGallery] = useState<boolean>(false)
+  const [photos, setPhotos] = useState<PhotoIdentifier[]>([])
+  const { permissionRequested, permissionGranted } = useGrantPermissions(
+    Platform.OS === 'ios' ? PERMISSIONS.IOS.MEDIA_LIBRARY : PERMISSIONS.ANDROID.READ_MEDIA_IMAGES
```
Comment (f1sh1918) [inline src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:61]: requesting android permissions for media is quite more complicated depending on versions
```js
const androidMediaPermission = Platform.Version >= 33 ? PERMISSIONS.ANDROID.READ_MEDIA_IMAGES : PERMISSIONS.ANDROID.READ_EXTERNAL_STORAGE;
```
For Android13 you will also need another permission

On Android 13 the READ_EXTERNAL_STORAGE has been [replace](https://developer.android.com/about/versions/13/behavior-changes-13#granular-media-permissions) by READ_MEDIA_IMAGES and READ_MEDIA_VIDEO.
```js
<manifest>
...
<uses-permission android:name="android.permission.READ_MEDIA_IMAGES"/>
<uses-permission android:name="android.permission.READ_MEDIA_VIDEO"/>
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
...
<application>
```
Also check for
```js
<application
      android:requestLegacyExternalStorage="true"
      ```

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 3]
```diff
@@ -33,39 +42,92 @@ const Container = styled.View`
...
+  )
+
+  const getGalleryPictures = () => {
+    CameraRoll.getPhotos({ first: 30, assetType: 'Photos' })
```
Comment (f1sh1918) [inline src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:65]: I think the number should be increased

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 4]
```diff
@@ -33,39 +42,92 @@ const Container = styled.View`
...
+          {permissionRequested && !permissionGranted && (
+            <NotAuthorisedView
+              description={getLabels().general.library.noAuthorization.description}
+              setVisible={setVisible}
+            />
+          )}
+          {permissionGranted && (
+            <ScrollView contentContainerStyle={{ flexDirection: 'row', flexWrap: 'wrap' }}>
+              {photos.map(photo => (
+                <Pressable key={photo.node.timestamp} onPress={() => selectImage(photo.node.image.uri)}>
```
Comment (f1sh1918) [inline src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:110]: when i increased amount of images in the gallery, i got an duplicated key error. Not sure whats the reason for. Timestamps should be unique

**Comment by LeandraH** [comment]
> Any reason why you didn't choose most popular library: https://www.npmjs.com/package/react-native-image-picker Just saw it could also take images. Not sure if it could maybe replace also the camera lib

Yes, I decided to go with the library that is recommended in the RN docs (https://reactnative.dev/docs/0.60/cameraroll). But I can switch it if you feel strongly about it. Or you can drop that into the ticket for replacing react-native-camera (https://issues.tuerantuer.org/browse/LUN-439).

**Comment by f1sh1918** [comment]
> > Any reason why you didn't choose most popular library: https://www.npmjs.com/package/react-native-image-picker Just saw it could also take images. Not sure if it could maybe replace also the camera lib
> 
> Yes, I decided to go with the library that is recommended in the RN docs (https://reactnative.dev/docs/0.60/cameraroll). But I can switch it if you feel strongly about it. Or you can drop that into the ticket for replacing react-native-camera (https://issues.tuerantuer.org/browse/LUN-439).

Well actually it's your ticket. So it was just a recomendación. I just think if we use one highly supported lib for two features it's better to maintain than two different libs. We could discuss this at the conference

**Comment by SteffiStuffel** [comment]
> 
> I just think if we use one highly supported lib for two features it's better to maintain than two different libs. We could discuss this at the conference

I think about this like Andy, but i am also fine with merging this now, change it later in the other ticket, although I think it would make sense to change it now, as then we/martin only need to test this one time and so on. Let me know how you decided and if i should review this.

**Comment by LeandraH** [comment]
Okay, I'll switch to the image-picker then, thanks for your input :)

**Comment by LeandraH** [comment]
I changed the package to react-native-image-crop-picker now. react-native-image-picker unfortunately can't handle the iOS' limited permission (https://github.com/react-native-image-picker/react-native-image-picker/issues/1749) but the crop-picker also has a camera function, so we can probably replace react-native-camera with it as well.

File: release-notes/unreleased/LUN-440-select-image-from-gallery.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+issue_key: LUN-440
+show_in_stores: true
+platforms:
+  - android
+  - ios
+de: Nutzer:innen können nun Bilder für eigene Vokablen aus der Fotomediathek benutzen.
```
Comment (SteffiStuffel) [inline release-notes/unreleased/LUN-440-select-image-from-gallery.yml:6]: Please delete these release notes, there will be one release note for the whole own vocabulary feature.

**Review by SteffiStuffel — COMMENTED** [review]
Tested on Android emulator. Please squash the commits.

**Review by SteffiStuffel — APPROVED** [review]
Nice work. Please squash the commits.

File: src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx [hunk 6]
```diff
@@ -50,17 +58,54 @@ const ImageSelectionOverlay = ({ setVisible, pushImage }: ImageSelectionOverlayP
...
+          reportError(e)
+        }
+      })
+      .finally(() => {
+        setIsGalleryOpen(false)
+      })
+  }
+
+  const androidMediaPermission =
+    // eslint-disable-next-line no-magic-numbers
```
Comment (f1sh1918) [inline src/routes/process-user-vocabulary/components/ImageSelectionOverlay.tsx:83]: please set a constant for the platform version and remove this :)

**Review by f1sh1918 — APPROVED** [review]
Tested on android and ios!
Nicely done :)

**Pull request merged**